Adds compilation support for Raspberry Pi

rotv opened the pull request:
A couple of years ago the https://github.com/ninjablocks/433Utils project copied files from rc-switch. Since then rc-switch has been updated and supports more remotes whereas the 433Utils copy has remained more or less unchanged.

To prevent this from keep happening, and to "gather forces" for development, I've incorporated rc-switch as a git submodule (not yet "pull requested"). For this to work the rc-switch project need some minor adjustments to support Raspberry Pi.

I've build and tested this on a Raspberry Pi and confirmed that it works. I don't have an Arduino so I can't test that.

NicoHood commented:
Why do we need this if there is a working raspberry pi build available over here?
https://github.com/arduino/Arduino/pull/3549#issuecomment-161739020

rotv commented:
rc-switch doesn't *need* this, but by adding these few lines the 433Utils project can directly compile against rc-switch which would be beneficial for rc-switch since more people would use the same code and combine efforts when it comes to supporting different devices.

I looked at the link you referred looks like it retrieves and compiles and entire IDE to be able to program Arduinos from Raspberry Pi? This sounds way heavier than what is needed since Arduinos are not part of the use case.

NicoHood commented:
You mean this is to use rc switch with the raspberry pi gpios?

rotv commented:
Exactly. I've just successfully turned on/off a light (socket) using 433Utils's RFSniffer and codesend compiled directly with the rotv/rc-switch code (both include the RCSwitch class).

NicoHood commented:
Okay looks like a nice idea :)
Sorry for missunderstanding.

rotv commented:
Now this PR is again ready to be merged. I've tested it with my Raspberry Pi and my protocol 1 remote and sockets.

Unfortunately I don't have the ability to test build for Arduino.

fingolfin commented on RCSwitch.h:
```diff
@@ -32,7 +33,29 @@
     #include "Arduino.h"
 #elif defined(ENERGIA) // LaunchPad, FraunchPad and StellarPad specific
-    #include "Energia.h"
+    #include "Energia.h"	
```
Your adding back a trailing space here which we removed recently.

rotv replied:
Ah, missed that. I've fixed it now.

fingolfin commented on .gitignore:
```diff
@@ -0,0 +1,17 @@
+*.lai
+*.la
+*.a
+
```
Why is this .gitignore being added? That is not needed to align the source code, is it? And it is not needed for the Arduino project either, is it?

rotv replied:
This is to prevent anyone from committing compiled files that would otherwise cause merge issues, I' m not sure which files are created during Arduino compilations but this file covers the normal C/C++ ones.

DS_Store is Mac OS X specific (folder metadata) but it won't harm anything, on the contrary it might be worth adding more stuff for different OS:es.

rotv replied:
Please let me know if this is what's preventing a merge.

fingolfin replied:
I don't think that it does (though I still don't understand why it would be necessary -- none of these file types are created by an Arduino build, as the Arduino app puts them into temp directories elsewhere on the system...

rotv replied:
Would you prefer if I removed it from the PR?

fingolfin commented:
Commented on some minor nitpicks, but overall, this seems fine to me. Thanks!

Matzz commented:
Porting to RPI is a great idea. Althought, in future you could look at http://wiringx.org/ to support more platforms.

rotv commented:
@Matzz, Nice advice on wiringx, hadn't seen that.

rotv commented:
ping @fingolfin

meGAmeS1 commented:
(ping @fingolfin)

Hey everyone, I'm so glad there are still active devs on this project. I recently bought a Raspberry Pi in order to control my remote control outlets. My first try was with 433Utils project, but sadly my outlets appear to use a rolling code. And luckily, I found these pull-request (#4, and this one #16) because my outlets are actually Brennenstuhl RC 3600. So I would be so glad to see both of these PR merged into the project and test if it actually works (I will also try to test on an Arduino, but I'm still waiting for the delivery).

I may have time for myself on the upcoming weekend, so I may try to merge both PR on my side if you can't merge them until then.

Anyway, thank you again, if I can help on anything feel free to ask!

fingolfin commented:
@meGAmeS1 I just closed #4 -- most of it was obsolete, and the remains were not in a mergable state, plus the original submitter never replied. And I don't think it would have helped at all with your problems, whatever they are, as the remaining code was simply an example.

I.e. there is nothing really to merge there -- only thing you could have done with PR #4 is to copy & paste the codes he used into your own program.

fingolfin commented on RCSwitch.h:
```diff
@@ -33,6 +34,28 @@
+extern "C"{
+#endif
+typedef uint8_t boolean;
+typedef uint8_t byte;
```
What are these two typedefs for?

fingolfin commented:
@rotv Note: the project owner is @sui77 not me, I just help out, so it might be better to ping him than me? :)

Anyway, I noticed one more tiny thing (what are those two typedefs for), but neither these nor the weird gitignore file are a reason from my POV to not merge this.

In other words: I am fine with merging this as it is (though if you could remove the typedefs, should they turn out to be unneeded, I"d consider that "cleaner").

And I am sorry you have to wait so long for this to be merged, but I did not realize that anybody was waiting for *me*, given that this is not really my project ... ;-)

So, if @sui77 does not complain, I'll assume he is fine with this to be merged, and would do it soon (in a couple days or so, to give him time to react)

rotv commented:
No worries @fingolfin. I think @sui77 hadn't been active for a while when I first sent in this PR, that's why I pinged you instead :)
The typedefs are needed but I cleaned up the indentation and added a comment as explanation.

rotv commented:
@meGAmeS1, You can have a look at https://github.com/rotv/433Utils/tree/rcswitch. I'll update it as soon as this pulse request is merged.

sui77 commented:
Sorry for my lack of responsiveness. As I currently don't have the time to deal in detail with this and I'm not keen enough to merge anything without actually verifying it... 

No, I don't I'm complain and I'm absolutely fine with you to do it :)

rotv commented:
You're absolutely right, works fine without and is much cleaner. Fixed in PR #48.
